docs/source/modeling_with_aequilibrae/route_choice/choice_set_generation.rst

```diff
@@ -0,0 +1,124 @@
+in question is computed. After each iteration, however, a pre-defined penalty factor is applied to all links that are
+part of the path found, essentially modifying the graph to make the previously found path less attractive.
+
+The LP method is a simple and effective way to generate route choice sets, but it is sensitive to the penalty factor,
+which can significantly affect the quality of the generated choice sets, requiring experimentation during the model
+development/estimation stage.
+
+The overhead of the LP method is negligible due to AequilibraE's internal data structures that allow for easy
+data manipulation of the graph in memory.
+
+
+BFS-LE
+~~~~~~
+
+At a high level, BFS-LE operates on a graph of graphs, exploring unique graphs linked by a single removed edge.
+Each graph can be uniquely categorised by a set of removed links from a common base graph, allowing us to avoid
+explicitly maintaining the graph of graphs. Instead, generating and storing that graph's set of removed links in the
+breadth-first search (BFS) order.
+
+To efficiently store and determine the uniqueness of a new route or removed link sets, we used modified hash functions
+with properties that allowed us to store and nest them within standard C++ data structures. We used a commutative hash
+function for the removed link sets to allow for amortised O(1) order-independent uniqueness testing. While the removed
+link sets are always constructed incrementally, we did not opt for an incremental hash function as we did not deem this
+a worthwhile optimisation. The removed link sets rarely grew larger than double digits, even on a network with over
+600,000 directed links. This may be an area worth exploring for networks with a significantly larger number of desired
+routes than links between ODs.
+
+For uniqueness testing of discovered routes, AequilibraE implements a traditional, non-commutative hash function.
+Since cryptographic security was not a requirement for our purposes, we use a fast general-purpose integer hash function.
+Further research could explore the use of specialised integer vector hash functions. As we did not find the hashing
+had a non-negligible influence on the runtime performance, this optimisation was not tested.
+
+AequilibraE also implements a combination of LP and BFS-LP as an optional feature to the latter algorithm, as
+recommended by Rieser-Schüssler et al. [1]_, which is also a reference for further details on the BFS-LE algorithm.
+
+Experiment
+----------
+
+In an experiment of with nearly 9,000 observed vehicle GPS routes covering a large Australian State, we found that
+all three algorithms (LP, BFS-LE, and BFS-LE+LP) had excellent performance in reproducing the observed routes. However,
+the computational overhead of BFS-LE is substantial enough to recommend always verifying if LP is fit-for-purpose.
+
+.. image:: ../../images/route_choice_sets.png
```

Comment:
@Jake-Moss , maybe change this picture to include all three (LP, BFS-LE, BFS-LE+LP at 1%)